### ISSUE TITLE
Request:  Automatically follow nextLink in continuation.

### ISSUE THREAD

**Issue by jamesra** [object/Simple.OData.Client#344]
I am using client.FindEntriesAsync to call functions which return multiple pages of results.

From the [issue 46](https://github.com/object/Simple.OData.Client/issues/46) on continuations I've learned to fetch the next link I must add a loop after every call which uses the NextPageLink property of the ODataFeedAnnotations object just in case I make a call which exceeds the results per page limit.

In a future version would it be possible to please have a client configuration option to automatically fetch all pages of a result set and return a single result?  Or perhaps default the client to collecting all results and add an option to limit the number of continuations followed for more advanced users?  That way I could simplify to a single line for each call again.

If you feel this is a straightforward change I may be willing to take it on.

Thanks!

**Comment by object** [comment]
Hi,

Thank you for the suggestion. I am not sure I want the library to take control over server-side control of such details. nextLink continuation has been added to OData to allow servers control the size of response payload, and I definitely don't want to make default behavior set to fetch all results. What if there are millions? Server may know this, client might not.

Another things is that I don't want the client to be unpredictable in how many network trips it makes to perform an operation. The library is called "Simple", the goal is to add a simple wrapper around OData HTTP requests. And it looks like in the suggested scenario the library will need to silently perform multiple HTTP calls that will look like a single one. This breaks current library approach.

Perhaps I am overlooking something but I am not aware of a method to overrule nextLink and request all results in a single HTTP call if server returns results split in pages.

**Comment by jamesra** [comment]
The multiple network request for a single call argument hadn't occurred to me.  It doesn't matter for my application but others may care.

I'm not convinced the client auto-following nextLink takes over the server-side control.  The server should specify max skips to prevent a client from fetching millions of records if that is an issue.

As an inexperienced API user I was getting incomplete result sets due to paging, with no indication the results were limited.  Adding ODataFeedAnnotations to the [documentation for paging](https://github.com/object/Simple.OData.Client/wiki/Results-projection,-paging-and-ordering) might help others in the future.  I apologize if it is documented somewhere that I missed. 

Thanks for your time.  Feel free to close the issue.

**Comment by NetTecture** [comment]
@object  I am not sur I buy your argument, WHile you are right in some cases, there are others where you are not. I may from the client side want to get all transactions of the month of november, EVEN IF I need multiple pages. There are arguments in both cases. A simple way to automatically dissolve server side paging into an enumeration with a specific limitation client side... is not necessarily bad.

**Comment by julealgon** [comment]
I agree with other posters here in that having an abstraction in the client library would work best with this.

`Microsoft.OData.Client`'s `DataServiceContext` has this in the form of the `GetAllPages()` method, which seems to work in a seamless way:
```csharp
var context = new DataServiceContext("http://localhost/api");
var results = await context
    .CreateQuery<EntityType>("entities")
    .GetAllPagesAsync();
```

I'd really appreciate if you could reconsider your initial decision on this one @object and consider adding a new method to fetch all results.

Notice how this doesn't need to be baked into an existing overload. It can be a new method on the `ODataClient` class that makes it clear pagination will take place. As long as you note in the method summary that you are potentially making multiple sequential calls to the OData API, I don't see how this violates any aspect of the library.

Also, keep in mind that "simple" also means making the life of the consumers simple enough. By not providing this support, you force consumers to implement significant logic to iterate on the pages.
